### DISCUSSION TITLE
River on k8s?

### DISCUSSION THREAD

**Discussion by mesquita** [riverqueue/river#216] [Q&A] [answered]
Hey hey,
what your experience running river on k8s?
My team has been experimenting using it in our service, the only problem that we seem to have is that the jobs are running on more than 1 pod at the same time. I don't know if it's a configuration problem on our side, but we thought we'd be fine due to the way river works. 

Thank you.
btw, we love river

**Comment by bgentry** [comment]
Are you saying that _an individual job_ is running on more than one pod at a time? That should not be possible, and we would need a lot more info about what you’re seeing to be able to look into it.

Or are you saying that each of your pods is running _different_ jobs? Because that’s how it’s supposed to work—each client is supposed to work jobs for whichever queues are specified in its config. You are supposed to run as many workers on the same queue as you want in order to distribute many jobs across them.

I don’t think there’s anything k8s specific to pay attention to with config. Each client is essentially stateless except for actively running jobs. As long as you give the Client time to shut down cleanly before terminating your process and respect context cancellation in workers you should be fine. You can utilize the more aggressive shutdown mode as needed if you really want to interrupt existing jobs immediately.

**Comment by gabrielclimb** [comment]
Hey @bgentry , I work together with @mesquita.
So, following this [article](https://www.cybertec-postgresql.com/en/tracking-changes-in-postgresql/), I was able to monitor the _river_leader_ table, it's changing a lot.
logs -> [leader_election_monitoring.csv](https://github.com/riverqueue/river/files/14334886/leader_election_monitoring.csv)
Not sure if suppose to work like this but i think this might be the root of our problems. 
might also related with #197

**Reply by bgentry** [reply]
Can you confirm if you’re on v0.0.20 which includes that bug fix?

**Reply by gabrielclimb** [reply]
I was in v0.0.19, just updated to v0.0.20. 
But now I'm getting this error:
`time=2024-02-19T17:14:09.231Z level=ERROR msg="error keeping leadership" elector.err="lost leadership with no error"`

I'm trying investigated why this happening aswell

**Reply by gabrielclimb** [reply]
![image](https://github.com/riverqueue/river/assets/21311863/f6d1bfb4-e7df-4c87-a52b-d18d8c82fb82)
it's going in a error loop
by the way, it should run every 2min

**Reply by bgentry** [reply]
@gabrielclimb I think you've hit a bug here, @brandur and I are confirming it now and working on a fix.

**Reply by gabrielclimb** [reply]
Let me know if there is a way I can help you guys.

**Reply by bgentry** [reply]
This should do it: https://github.com/riverqueue/river/pull/217

In the run up to release, we overlooked adding comprehensive test coverage on leader election, but it's pretty clear we need that. We'll get that added soon, sorry for the trouble!

**Reply by bgentry** [reply, marked as answer]
@gabrielclimb We will have that out in a release shortly. However this issue should only have resulted in duplicated work for the responsibilities of the leader such as pruning old jobs, inserting scheduled jobs, etc. If you have any reason to believe the same job was being worked twice simultaneously, please do share more details as that absolutely should not happen :pray: Thanks for all the info to help us fix this fast.

**Reply by gabrielclimb** [reply]
Thank you @bgentry , really appreciated

**Reply by brandur** [reply]
@gabrielclimb We just cut https://github.com/riverqueue/river/releases/tag/v0.0.21 containing a bug fix for this. Please give it a shot and let us know how it goes.

**Reply by gabrielclimb** [reply]
just updated, looks like it is fixed.
![image](https://github.com/riverqueue/river/assets/21311863/7c86d0b2-b8e8-417a-be5e-0a4fc25b8401)


thank you very much guys
@bgentry @brandur

**Reply by mesquita** [reply]
Thank you so much!! @bgentry @brandur @gabrielclimb

**Reply by brandur** [reply]
Hey @gabrielclimb @mesquita, actually we found one more related bug in the elector and pushed another fix for it in https://github.com/riverqueue/river/releases/tag/v0.0.22, and you guys should update to that if you can. Sorry about the churn (we think that's the last one).

**Reply by gabrielclimb** [reply]
@brandur tks to let us know. ❤️
